src/Adapter/File/Uploader/AttachmentFileUploader.php

```
@@ -0,0 +1,125 @@
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\File\Uploader;
+
+use Attachment;
+use PrestaShop\PrestaShop\Adapter\Configuration;
+use PrestaShop\PrestaShop\Core\Domain\Attachment\Exception\AttachmentConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Attachment\Exception\AttachmentNotFoundException;
+use PrestaShop\PrestaShop\Core\Domain\Attachment\Exception\AttachmentUploadFailedException;
+use PrestaShopException;
+use Symfony\Component\HttpFoundation\File\Exception\FileException;
+use Tools;
+
+/**
+ * Uploads attachment file and if needed deletes old attachment file
+ *
+ * @internal
+ */
+final class AttachmentFileUploader
+{
+    /**
+     * @param string $filePath
+     * @param string $uniqueFileName
+     * @param int $fileSize
+     * @param int|null $id
+     *
+     * @throws AttachmentConstraintException
+     * @throws AttachmentNotFoundException
+     * @throws AttachmentUploadFailedException
+     */
+    public function upload(string $filePath, string $uniqueFileName, int $fileSize, ?int $id = null)
+    {
+        $this->checkFileAllowedForUpload($fileSize);
+        $this->uploadFile($filePath, $uniqueFileName, $fileSize);
+        if ($id !== null) {
+            $this->deleteOldFile($id);
+        }
+    }
+
+    /**
+     * @param int $attachmentId
+     *
+     * @throws AttachmentNotFoundException
+     */
+    private function deleteOldFile(int $attachmentId)
+    {
+        try {
+            $attachment = new Attachment($attachmentId);
+            @unlink(_PS_DOWNLOAD_DIR_ . $attachment->file);
```

Let's remove this `@` that comes from legacy. If the unlink failed we want to know why.

I think it was put here in order to avoid errors if `deleteOldFile` was used on a bad file (the file has been removed directly on the server then the database record is removed).
If it is indeed the reason for this `@` then let's use a flag `throwExceptionIfFileDoesNotExist` and catch the error. If flag is true we throw the error. If flag is false we silence it. But at least it's intentional.

Right now the unlink could fail for multiple reasons (bad filepath, bad database content, bad read/write settings) and the user will never know.

---

I think this `@` was only used for tmp directory file unlink and elsewhere it wasn't present. So because symfony transforms warnings into ContextErrorException, i could display the message in controller level `'Unexpected error occurred: ' . $e->getMessage` and remove `@` altogether. For the flag if i understand correctly, `deleteOldFile ` method should accept this parameter and throw `FileNotFoundException` in case flag is `true` and file doesn't exist and suppress errors otherwise with `@unlink`?

---

> I think this `@` was only used for tmp directory file unlink and elsewhere it wasn't present. So because symfony transforms warnings into ContextErrorException, i could display the message in controller level `'Unexpected error occurred: ' . $e->getMessage` and remove `@` altogether.

This works 👍 

> For the flag if i understand correctly, `deleteOldFile ` method should accept this parameter and throw `FileNotFoundException` in case flag is `true` and file doesn't exist and suppress errors otherwise with `@unlink`?

Yes, the flag should control whether we put the `@` or not. We can imagine someone wanting to delete a bulk list of 100 files and so he does not want 30 warning messages if 30 files are corrupted.